File: backend/routes/analytics.js
const express = require('express');
const Joi = require('joi');
const TestResult = require('../models/TestResult');
const { dynamodb, TABLES } = require('../config/dynamodb');

const router = express.Router();

// Validation schemas
const analyticsQuerySchema = Joi.object({
  startDate: Joi.string().isoDate().optional(),
  endDate: Joi.string().isoDate().optional(),
  testType: Joi.string().valid('quickTest', 'detailedAnalysis', 'manualTest').optional(),
  groupBy: Joi.string().valid('day', 'week', 'month', 'testType', 'location').optional(),
  limit: Joi.number().integer().min(1).max(1000).optional()
});

// GET /api/analytics/performance - Get performance analytics
router.get('/performance', async (req, res, next) => {
  try {
    const { error, value } = analyticsQuerySchema.validate(req.query);
    if (error) {
      return res.status(400).json({
        error: 'Validation error',
        details: error.details.map(detail => detail.message)
      });
    }

    const { startDate, endDate, testType, limit = 100 } = value;
    
    // Build filters
    const filters = {};
    if (startDate && endDate) {
      filters.startDate = startDate;
      filters.endDate = endDate;
    }
    if (testType) {
      filters.testType = testType;
    }

    const results = await TestResult.getWithFilters(filters, limit);

    // Calculate performance metrics
    const performanceData = {
      downloadSpeeds: [],
      uploadSpeeds: [],
      latencies: [],
      connectionQualities: {},
      testTypeDistribution: {},
      timeSeriesData: {},
      summary: {
        totalTests: results.length,
        averageDownloadSpeed: 0,
        averageUploadSpeed: 0,
        averageLatency: 0,
        bestDownloadSpeed: 0,
        bestUploadSpeed: 0,
        lowestLatency: Infinity
      }
    };

    let totalDownload = 0;
    let totalUpload = 0;
    let totalLatency = 0;
    let speedCount = 0;
    let latencyCount = 0;

    results.forEach(result => {
      // Count test types
      performanceData.testTypeDistribution[result.testType] = 
        (performanceData.testTypeDistribution[result.testType] || 0) + 1;

      if (result.networkData && result.networkData.speedTest) {
        const download = parseFloat(result.networkData.speedTest.download);
        const upload = parseFloat(result.networkData.speedTest.upload);
        const latency = result.networkData.speedTest.latency;
        const quality = result.networkData.speedTest.connectionQuality;

        if (!isNaN(download)) {
          performanceData.downloadSpeeds.push(download);
          totalDownload += download;
          speedCount++;
          performanceData.summary.bestDownloadSpeed = Math.max(performanceData.summary.bestDownloadSpeed, download);
        }

        if (!isNaN(upload)) {
          performanceData.uploadSpeeds.push(upload);
          totalUpload += upload;
          performanceData.summary.bestUploadSpeed = Math.max(performanceData.summary.bestUploadSpeed, upload);
        }

        if (latency && !isNaN(latency)) {
          performanceData.latencies.push(latency);
          totalLatency += latency;
          latencyCount++;
          performanceData.summary.lowestLatency = Math.min(performanceData.summary.lowestLatency, latency);
        }

        if (quality) {
          performanceData.connectionQualities[quality] = 
            (performanceData.connectionQualities[quality] || 0) + 1;
        }
      }

      // Time series data
      const testDate = new Date(result.timestamp);
      const dateKey = testDate.toISOString().split('T')[0];
      if (!performanceData.timeSeriesData[dateKey]) {
        performanceData.timeSeriesData[dateKey] = {
          tests: 0,
          avgDownload: 0,
          avgUpload: 0,
          avgLatency: 0
        };
      }
      performanceData.timeSeriesData[dateKey].tests++;
    });

    // Calculate averages
    if (speedCount > 0) {
      performanceData.summary.averageDownloadSpeed = (totalDownload / speedCount).toFixed(2);
      performanceData.summary.averageUploadSpeed = (totalUpload / speedCount).toFixed(2);
    }
    if (latencyCount > 0) {
      performanceData.summary.averageLatency = (totalLatency / latencyCount).toFixed(2);
    }

    // Sort time series data
    performanceData.timeSeriesData = Object.entries(performanceData.timeSeriesData)
      .sort(([a], [b]) => a.localeCompare(b))
      .reduce((obj, [key, value]) => {
        obj[key] = value;
        return obj;
      }, {});

    res.json({
      success: true,
      data: performanceData
    });

  } catch (error) {
    next(error);
  }
});

// GET /api/analytics/trends - Get trend analysis
router.get('/trends', async (req, res, next) => {
  try {
    const { error, value } = analyticsQuerySchema.validate(req.query);
    if (error) {
      return res.status(400).json({
        error: 'Validation error',
        details: error.details.map(detail => detail.message)
      });
    }

    const { startDate, endDate, limit = 200 } = value;
    
    const filters = {};
    if (startDate && endDate) {
      filters.startDate = startDate;
      filters.endDate = endDate;
    }

    const results = await TestResult.getWithFilters(filters, limit);

    // Group by time periods
    const trends = {
      daily: {},
      weekly: {},
      monthly: {},
      testTypeTrends: {},
      performanceTrends: {
        download: {},
        upload: {},
        latency: {}
      }
    };

    results.forEach(result => {
      const testDate = new Date(result.timestamp);
      const dayKey = testDate.toISOString().split('T')[0];
      const weekKey = getWeekKey(testDate);
      const monthKey = testDate.toISOString().slice(0, 7); // YYYY-MM

      // Daily trends
      if (!trends.daily[dayKey]) {
        trends.daily[dayKey] = {
          tests: 0,
          downloadSpeeds: [],
          uploadSpeeds: [],
          latencies: []
        };
      }
      trends.daily[dayKey].tests++;

      // Weekly trends
      if (!trends.weekly[weekKey]) {
        trends.weekly[weekKey] = {
          tests: 0,
          downloadSpeeds: [],
          uploadSpeeds: [],
          latencies: []
        };
      }
      trends.weekly[weekKey].tests++;

      // Monthly trends
      if (!trends.monthly[monthKey]) {
        trends.monthly[monthKey] = {
          tests: 0,
          downloadSpeeds: [],
          uploadSpeeds: [],
          latencies: []
        };
      }
      trends.monthly[monthKey].tests++;

      // Test type trends
      if (!trends.testTypeTrends[result.testType]) {
        trends.testTypeTrends[result.testType] = {
          daily: {},
          weekly: {},
          monthly: {}
        };
      }
      trends.testTypeTrends[result.testType].daily[dayKey] = 
        (trends.testTypeTrends[result.testType].daily[dayKey] || 0) + 1;

      // Performance data
      if (result.networkData && result.networkData.speedTest) {
        const download = parseFloat(result.networkData.speedTest.download);
        const upload = parseFloat(result.networkData.speedTest.upload);
        const latency = result.networkData.speedTest.latency;

        if (!isNaN(download)) {
          trends.daily[dayKey].downloadSpeeds.push(download);
          trends.weekly[weekKey].downloadSpeeds.push(download);
          trends.monthly[monthKey].downloadSpeeds.push(download);
        }

        if (!isNaN(upload)) {
          trends.daily[dayKey].uploadSpeeds.push(upload);
          trends.weekly[weekKey].uploadSpeeds.push(upload);
          trends.monthly[monthKey].uploadSpeeds.push(upload);
        }

        if (latency && !isNaN(latency)) {
          trends.daily[dayKey].latencies.push(latency);
          trends.weekly[weekKey].latencies.push(latency);
          trends.monthly[monthKey].latencies.push(latency);
        }
      }
    });

    // Calculate averages for each period
    Object.keys(trends.daily).forEach(day => {
      const dayData = trends.daily[day];
      dayData.avgDownload = dayData.downloadSpeeds.length > 0 
        ? (dayData.downloadSpeeds.reduce((a, b) => a + b, 0) / dayData.downloadSpeeds.length).toFixed(2)
        : 0;
      dayData.avgUpload = dayData.uploadSpeeds.length > 0
        ? (dayData.uploadSpeeds.reduce((a, b) => a + b, 0) / dayData.uploadSpeeds.length).toFixed(2)
        : 0;
      dayData.avgLatency = dayData.latencies.length > 0
        ? (dayData.latencies.reduce((a, b) => a + b, 0) / dayData.latencies.length).toFixed(2)
        : 0;
    });

    res.json({
      success: true,
      data: trends
    });

  } catch (error) {
    next(error);
  }
});

// GET /api/analytics/comparison - Compare performance across different criteria
router.get('/comparison', async (req, res, next) => {
  try {
    const { error, value } = analyticsQuerySchema.validate(req.query);
    if (error) {
      return res.status(400).json({
        error: 'Validation error',
        details: error.details.map(detail => detail.message)
      });
    }

    const { startDate, endDate, limit = 500 } = value;
    
    const filters = {};
    if (startDate && endDate) {
      filters.startDate = startDate;
      filters.endDate = endDate;
    }

    const results = await TestResult.getWithFilters(filters, limit);

    const comparison = {
      testTypes: {},
      timeOfDay: {},
      dayOfWeek: {},
      locations: {},
      devices: {}
    };

    results.forEach(result => {
      const testDate = new Date(result.timestamp);
      const hour = testDate.getHours();
      const dayOfWeek = testDate.toDayString();
      const location = result.ipAddress || 'Unknown';
      const device = result.userAgent ? getDeviceType(result.userAgent) : 'Unknown';

      // Test type comparison
      if (!comparison.testTypes[result.testType]) {
        comparison.testTypes[result.testType] = {
          count: 0,
          downloadSpeeds: [],
          uploadSpeeds: [],
          latencies: []
        };
      }
      comparison.testTypes[result.testType].count++;

      // Time of day comparison
      const timeSlot = getTimeSlot(hour);
      if (!comparison.timeOfDay[timeSlot]) {
        comparison.timeOfDay[timeSlot] = {
          count: 0,
          downloadSpeeds: [],
          uploadSpeeds: [],
          latencies: []
        };
      }
      comparison.timeOfDay[timeSlot].count++;

      // Day of week comparison
      if (!comparison.dayOfWeek[dayOfWeek]) {
        comparison.dayOfWeek[dayOfWeek] = {
          count: 0,
          downloadSpeeds: [],
          uploadSpeeds: [],
          latencies: []
        };
      }
      comparison.dayOfWeek[dayOfWeek].count++;

      // Location comparison
      if (!comparison.locations[location]) {
        comparison.locations[location] = {
          count: 0,
          downloadSpeeds: [],
          uploadSpeeds: [],
          latencies: []
        };
      }
      comparison.locations[location].count++;

      // Device comparison
      if (!comparison.devices[device]) {
        comparison.devices[device] = {
          count: 0,
          downloadSpeeds: [],
          uploadSpeeds: [],
          latencies: []
        };
      }
      comparison.devices[device].count++;

      // Add performance data
      if (result.networkData && result.networkData.speedTest) {
        const download = parseFloat(result.networkData.speedTest.download);
        const upload = parseFloat(result.networkData.speedTest.upload);
        const latency = result.networkData.speedTest.latency;

        if (!isNaN(download)) {
          comparison.testTypes[result.testType].downloadSpeeds.push(download);
          comparison.timeOfDay[timeSlot].downloadSpeeds.push(download);
          comparison.dayOfWeek[dayOfWeek].downloadSpeeds.push(download);
          comparison.locations[location].downloadSpeeds.push(download);
          comparison.devices[device].downloadSpeeds.push(download);
        }

        if (!isNaN(upload)) {
          comparison.testTypes[result.testType].uploadSpeeds.push(upload);
          comparison.timeOfDay[timeSlot].uploadSpeeds.push(upload);
          comparison.dayOfWeek[dayOfWeek].uploadSpeeds.push(upload);
          comparison.locations[location].uploadSpeeds.push(upload);
          comparison.devices[device].uploadSpeeds.push(upload);
        }

        if (latency && !isNaN(latency)) {
          comparison.testTypes[result.testType].latencies.push(latency);
          comparison.timeOfDay[timeSlot].latencies.push(latency);
          comparison.dayOfWeek[dayOfWeek].latencies.push(latency);
          comparison.locations[location].latencies.push(latency);
          comparison.devices[device].latencies.push(latency);
        }
      }
    });

    // Calculate averages for each category
    Object.keys(comparison).forEach(category => {
      Object.keys(comparison[category]).forEach(key => {
        const data = comparison[category][key];
        data.avgDownload = data.downloadSpeeds.length > 0
          ? (data.downloadSpeeds.reduce((a, b) => a + b, 0) / data.downloadSpeeds.length).toFixed(2)
          : 0;
        data.avgUpload = data.uploadSpeeds.length > 0
          ? (data.uploadSpeeds.reduce((a, b) => a + b, 0) / data.uploadSpeeds.length).toFixed(2)
          : 0;
        data.avgLatency = data.latencies.length > 0
          ? (data.latencies.reduce((a, b) => a + b, 0) / data.latencies.length).toFixed(2)
          : 0;
      });
    });

    res.json({
      success: true,
      data: comparison
    });

  } catch (error) {
    next(error);
  }
});

// Helper functions
function getWeekKey(date) {
  const year = date.getFullYear();
  const week = Math.ceil((date.getDate() + date.getDay()) / 7);
  return `${year}-W${week.toString().padStart(2, '0')}`;
}

function getTimeSlot(hour) {
  if (hour >= 6 && hour < 12) return 'Morning (6-12)';
  if (hour >= 12 && hour < 18) return 'Afternoon (12-18)';
  if (hour >= 18 && hour < 24) return 'Evening (18-24)';
  return 'Night (0-6)';
}

function getDeviceType(userAgent) {
  const ua = userAgent.toLowerCase();
  if (ua.includes('mobile') || ua.includes('android') || ua.includes('iphone')) {
    return 'Mobile';
  } else if (ua.includes('tablet') || ua.includes('ipad')) {
    return 'Tablet';
  } else {
    return 'Desktop';
  }
}

module.exports = router; 

File: backend-python/routes/analytics.py
"""
Analytics API Routes
"""

from flask import Blueprint, jsonify, request
from models.test_result import TestResult
from datetime import datetime, timedelta
from collections import defaultdict

analytics_bp = Blueprint('analytics', __name__)

# GET /api/analytics/performance - Get performance analytics
@analytics_bp.route('/performance', methods=['GET'])
def get_performance_analytics():
    """Get performance analytics"""
    try:
        # Get filters from query params
        start_date = request.args.get('startDate')
        end_date = request.args.get('endDate')
        test_type = request.args.get('testType')
        limit = int(request.args.get('limit', 100))
        
        # Build filters
        filters = {}
        if start_date and end_date:
            filters['startDate'] = start_date
            filters['endDate'] = end_date
        if test_type:
            filters['testType'] = test_type
        
        results = TestResult.get_with_filters(filters, limit)
        
        # Calculate performance metrics
        performance_data = {
            'downloadSpeeds': [],
            'uploadSpeeds': [],
            'latencies': [],
            'connectionQualities': {},
            'testTypeDistribution': {},
            'timeSeriesData': {},
            'summary': {
                'totalTests': len(results),
                'averageDownloadSpeed': 0,
                'averageUploadSpeed': 0,
                'averageLatency': 0,
                'bestDownloadSpeed': 0,
                'bestUploadSpeed': 0,
                'lowestLatency': float('inf')
            }
        }
        
        total_download = 0
        total_upload = 0
        total_latency = 0
        speed_count = 0
        latency_count = 0
        
        for result in results:
            # Count test types
            test_type = result.get('testType')
            performance_data['testTypeDistribution'][test_type] = \
                performance_data['testTypeDistribution'].get(test_type, 0) + 1
            
            network_data = result.get('networkData', {})
            speed_test = network_data.get('speedTest', {})
            
            download = float(speed_test.get('download', 0))
            upload = float(speed_test.get('upload', 0))
            latency = speed_test.get('latency')
            quality = speed_test.get('connectionQuality')
            
            if download:
                performance_data['downloadSpeeds'].append(download)
                total_download += download
                speed_count += 1
                performance_data['summary']['bestDownloadSpeed'] = max(
                    performance_data['summary']['bestDownloadSpeed'], download
                )
            
            if upload:
                performance_data['uploadSpeeds'].append(upload)
                total_upload += upload
                performance_data['summary']['bestUploadSpeed'] = max(
                    performance_data['summary']['bestUploadSpeed'], upload
                )
            
            if latency:
                performance_data['latencies'].append(latency)
                total_latency += latency
                latency_count += 1
                performance_data['summary']['lowestLatency'] = min(
                    performance_data['summary']['lowestLatency'], latency
                )
            
            if quality:
                performance_data['connectionQualities'][quality] = \
                    performance_data['connectionQualities'].get(quality, 0) + 1
            
            # Time series data
            timestamp = result.get('timestamp', '')
            if timestamp:
                date_key = timestamp.split('T')[0]
                if date_key not in performance_data['timeSeriesData']:
                    performance_data['timeSeriesData'][date_key] = {
                        'tests': 0,
                        'avgDownload': 0,
                        'avgUpload': 0,
                        'avgLatency': 0
                    }
                performance_data['timeSeriesData'][date_key]['tests'] += 1
        
        # Calculate averages
        if speed_count > 0:
            performance_data['summary']['averageDownloadSpeed'] = round(total_download / speed_count, 2)
            performance_data['summary']['averageUploadSpeed'] = round(total_upload / speed_count, 2)
        if latency_count > 0:
            performance_data['summary']['averageLatency'] = round(total_latency / latency_count, 2)
        
        return jsonify({
            'success': True,
            'data': performance_data
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': 'Internal server error',
            'message': str(e)
        }), 500

# GET /api/analytics/trends - Get trend analysis
@analytics_bp.route('/trends', methods=['GET'])
def get_trend_analytics():
    """Get trend analysis"""
    try:
        limit = int(request.args.get('limit', 200))
        results = TestResult.get_recent(limit)
        
        trends = {
            'daily': defaultdict(lambda: {'tests': 0, 'downloadSpeeds': [], 'uploadSpeeds': [], 'latencies': []}),
            'testTypeTrends': defaultdict(lambda: {'daily': defaultdict(int)})
        }
        
        for result in results:
            timestamp = result.get('timestamp', '')
            if not timestamp:
                continue
            
            date_key = timestamp.split('T')[0]
            test_type = result.get('testType')
            
            # Daily trends
            trends['daily'][date_key]['tests'] += 1
            
            # Test type trends
            trends['testTypeTrends'][test_type]['daily'][date_key] += 1
            
            # Performance data
            network_data = result.get('networkData', {})
            speed_test = network_data.get('speedTest', {})
            
            if speed_test.get('download'):
                trends['daily'][date_key]['downloadSpeeds'].append(float(speed_test['download']))
            if speed_test.get('upload'):
                trends['daily'][date_key]['uploadSpeeds'].append(float(speed_test['upload']))
            if speed_test.get('latency'):
                trends['daily'][date_key]['latencies'].append(speed_test['latency'])
        
        # Convert defaultdicts to regular dicts
        trends['daily'] = dict(trends['daily'])
        trends['testTypeTrends'] = {k: {'daily': dict(v['daily'])} for k, v in trends['testTypeTrends'].items()}
        
        return jsonify({
            'success': True,
            'data': trends
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': 'Internal server error',
            'message': str(e)
        }), 500

# GET /api/analytics/comparison - Compare performance across different criteria
@analytics_bp.route('/comparison', methods=['GET'])
def get_comparison_analytics():
    """Compare performance across different criteria"""
    try:
        limit = int(request.args.get('limit', 500))
        results = TestResult.get_recent(limit)
        
        comparison = {
            'testTypes': defaultdict(lambda: {'count': 0, 'downloadSpeeds': [], 'uploadSpeeds': [], 'latencies': []}),
            'timeOfDay': defaultdict(lambda: {'count': 0, 'downloadSpeeds': [], 'uploadSpeeds': [], 'latencies': []}),
            'dayOfWeek': defaultdict(lambda: {'count': 0, 'downloadSpeeds': [], 'uploadSpeeds': [], 'latencies': []})
        }
        
        for result in results:
            test_type = result.get('testType')
            timestamp = result.get('timestamp', '')
            
            if timestamp:
                try:
                    dt = datetime.fromisoformat(timestamp.replace('Z', '+00:00'))
                    hour = dt.hour
                    day_of_week = dt.strftime('%A')
                    
                    # Determine time slot
                    if 6 <= hour < 12:
                        time_slot = 'Morning (6-12)'
                    elif 12 <= hour < 18:
                        time_slot = 'Afternoon (12-18)'
                    elif 18 <= hour < 24:
                        time_slot = 'Evening (18-24)'
                    else:
                        time_slot = 'Night (0-6)'
                    
                    # Increment counts
                    comparison['testTypes'][test_type]['count'] += 1
                    comparison['timeOfDay'][time_slot]['count'] += 1
                    comparison['dayOfWeek'][day_of_week]['count'] += 1
                    
                    # Add performance data
                    network_data = result.get('networkData', {})
                    speed_test = network_data.get('speedTest', {})
                    
                    if speed_test.get('download'):
                        download = float(speed_test['download'])
                        comparison['testTypes'][test_type]['downloadSpeeds'].append(download)
                        comparison['timeOfDay'][time_slot]['downloadSpeeds'].append(download)
                        comparison['dayOfWeek'][day_of_week]['downloadSpeeds'].append(download)
                    
                    if speed_test.get('upload'):
                        upload = float(speed_test['upload'])
                        comparison['testTypes'][test_type]['uploadSpeeds'].append(upload)
                        comparison['timeOfDay'][time_slot]['uploadSpeeds'].append(upload)
                        comparison['dayOfWeek'][day_of_week]['uploadSpeeds'].append(upload)
                    
                    if speed_test.get('latency'):
                        latency = speed_test['latency']
                        comparison['testTypes'][test_type]['latencies'].append(latency)
                        comparison['timeOfDay'][time_slot]['latencies'].append(latency)
                        comparison['dayOfWeek'][day_of_week]['latencies'].append(latency)
                except:
                    pass
        
        # Calculate averages for each category
        for category in comparison:
            for key in comparison[category]:
                data = comparison[category][key]
                data['avgDownload'] = round(sum(data['downloadSpeeds']) / len(data['downloadSpeeds']), 2) if data['downloadSpeeds'] else 0
                data['avgUpload'] = round(sum(data['uploadSpeeds']) / len(data['uploadSpeeds']), 2) if data['uploadSpeeds'] else 0
                data['avgLatency'] = round(sum(data['latencies']) / len(data['latencies']), 2) if data['latencies'] else 0
        
        # Convert defaultdicts to regular dicts
        comparison = {k: dict(v) for k, v in comparison.items()}
        
        return jsonify({
            'success': True,
            'data': comparison
        }), 200
        
    except Exception as e:
        return jsonify({
            'error': 'Internal server error',
            'message': str(e)
        }), 500

